dask/tests/test_delayed.py

```diff
@@ -93,24 +94,132 @@ def test_delayed():
 
 
 def test_delayed_with_dataclass():
-    dataclasses = pytest.importorskip("dataclasses")
+    @dataclass()
+    class ADataClass:
+        a: int
 
-    # Avoid @dataclass decorator as Python < 3.7 fail to interpret the type hints
-    ADataClass = dataclasses.make_dataclass(
-        "ADataClass", [("a", int), ("b", int, dataclasses.field(init=False))]
-    )
+    literal = delayed(3)
+    with_class = delayed({"data": ADataClass(a=literal)})
+
+    def return_nested(obj):
+        return obj["data"].a
+
+    final = delayed(return_nested)(with_class)
+
+    assert final.compute() == 3
+
+
+def test_delayed_with_frozen_dataclass():
+    @dataclass(frozen=True)
+    class ADataClass:
+        a: int
+
+    literal = delayed(3)
+    with_class = delayed({"data": ADataClass(a=literal)})
+
+    def return_nested(obj):
+        return obj["data"].a
+
+    final = delayed(return_nested)(with_class)
+
+    assert final.compute() == 3
+
+
+def test_delayed_with_dataclass_with_custom_init():
+    @dataclass()
+    class ADataClass:
+        a: int
+
+        def __init__(self, b: int):
+            self.a = b
 
     literal = dask.delayed(3)
-    with_class = dask.delayed({"a": ADataClass(a=literal)})
+
+    with pytest.raises(TypeError) as e:
+        dask.delayed({"data": ADataClass(b=literal)})
+
```

Comment:
It would be great to also test that the original issue is fixed here (or in a separate test). That is, if we don't find a collection, don't bother traversing the dataclass and treat it as an opaque value.